### DISCUSSION TITLE
Use functions at the scripts ?

### DISCUSSION THREAD

**Discussion by CvH** [tteck/Proxmox#1198] [General]
Hi, not sure that was asked already but wouldn't make sense to use functions that are at some common file ?

Currently all scripts are more or less standalone and have a lot stuff over and over again - that makes it difficult to maintain and rather bloated.

So a idea would be to move all common stuff that is used over and over again in some function.sh that is downloaded and called at the start of the script. Would make the install scripts MUCH cleaner and the actually logic written one time instead ~100 times.

**Comment by tteck** [comment]
~The problem I'm facing is that I'm able to successfully source a functions file for the first script (-v5.sh), but I'm unable to do so for the third script (-v5-install.sh) that is invoked by the first script. I'm confident that there's a solution to this issue, but I'm not sure how to proceed.~

At present, there is a preliminary version of this in the next branch.

**Reply by tteck** [reply]
Now, I am in the process of testing after finally figuring it out.

**Reply by CvH** [reply]
that looks really good so far :) 
I would suggest (no idea what you have already planed) move the whole apt stuff to a function too so you can easily control the verbose of the output and you can add functionality to do a `install_package abc` instead of calling apt or apk etc ... .
So you can always use install_package abc instead of apt install abc, or pkg add abc .... . If you progress further you can even put the enable daemons stuff into a function and add a arch specific detection. You could end up with one installer for different distros (debian/alpine ...) while all overhead is a distro chooser at the beginning.

And one thing I recognized that may not be a problem for you but everyone else looking at the scripts, pls try to use better function names ;)

just a proposal
```
color                 ->  func_color
verb_ip6              ->  func_network_ipv6
catch_errors          ->  func_error_catcher
error_handler         ->  func_error_handler
msg_info              ->  func_msg_info
msg_ok                ->  func_msg_ok
msg_error             ->  func_msg_error
PVE_CHECK             ->  func_check_pve
ARCH_CHECK            ->  func_check_arch
setting_up_container  ->  func_create_container
network_check         ->  func_check_network
update_os             ->  func_update_os
motd_ssh              ->  func_set_motd_ssh
customize             ->  func_container_customize
description           ->  func_lxc_description
```

**Reply by tteck** [reply]
Thanks for bringing this up, as it prompted me to revisit the challenges of passing function files to scripts that are called by other scripts. While I did not utilize your suggested naming convention, your input is still valuable to me.

**Reply by CvH** [reply]
`I did not utilize your suggested naming convention` no problem, everyone has different habits :)
You have a lot indent mixings and some questionable bash formatting at some places but nothing too crazy after scrolling through the changes.
For example https://github.com/tteck/Proxmox/blob/next/misc/alpine.func#L199-L207
I would pr some cleanup after you switched your branches 👍 

Btw watch out for `Var=$Var2` (should be `Var="$Var2"`), that can easily expand totally wrong :)
https://github.com/tteck/Proxmox/blob/next/misc/alpine.func#L350-L358
https://github.com/tteck/Proxmox/blob/next/misc/alpine.func#L387

**Reply by tteck** [reply]
I'm not sure what caused the indentation issue, but I have since run shfmt to fix it.
